### PR TITLE
Get article comments

### PR DESCRIPTION
**Pull request by good4something-6**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -1,11 +1,13 @@
+const { describe, expect } = require("@jest/globals");
+const { test, beforeEach, afterAll } = require("@jest/globals");
```
Comment (HeatherEver) [inline __tests__/app.test.js]: Think these lines have auto imported - if they have just delete them before they cause issues!

File: __tests__/app.test.js [hunk 2]
```diff
@@ -145,6 +138,57 @@ describe("Articles", () => {
...
         });
     });
+    test("404 - article id is not in database", () => {
+      return request(app)
+        .patch("/api/articles/99999")
+        .send({ inc_votes: 1 })
+        .expect(404)
+        .then((result) => {
+          expect(result.body.msg).toBe("404 - Article Not Found");
+        });
```
Comment (HeatherEver) [inline __tests__/app.test.js:141]: it appears you have no status 200 tests for your PATCH request?

File: __tests__/app.test.js [hunk 3]
```diff
@@ -145,6 +138,57 @@ describe("Articles", () => {
...
+
+    test("404 article_id provided is not in database", () => {
+      return request(app)
+        .get("/api/articles/999/comments")
+        .expect(404)
+        .then((result) => {
+          expect(result.body.msg).toBe("404 - Article Not Found");
+        });
+    });
+    test("404 article_id provided is not valid type", () => {
```
Comment (HeatherEver) [inline __tests__/app.test.js:176]: Change the status code in the test description 😄

**Review by HeatherEver — COMMENTED** [review]
Looks really good! Just a couple of small changes to make!

File: __tests__/app.test.js [hunk 4]
```diff
@@ -1,11 +1,11 @@
 const request = require("supertest");
+require("jest-sorted");
```
Comment (dentednerd) [inline __tests__/app.test.js]: You don't need to require `jest-sorted` here.

File: __tests__/app.test.js [hunk 5]
```diff
@@ -127,6 +118,15 @@ describe("Articles", () => {
...
+    test("404 - article id is not in database", () => {
+      return request(app)
+        .patch("/api/articles/99999")
+        .send({ inc_votes: 1 })
+        .expect(404)
+        .then((result) => {
+          expect(result.body.msg).toBe("404 - Article Not Found");
+        });
+    });
     test("200 - valid article id is provided and votes updates by 1 correctly", () => {
```
Comment (dentednerd) [inline __tests__/app.test.js]: Conventionally, we put happy path tests above sad path tests!

File: __tests__/app.test.js [hunk 6]
```diff
@@ -156,10 +198,8 @@ describe("Articles", () => {
           const articles = result.body;
           expect(Array.isArray(articles)).toBe(true);
           expect(articles).toHaveLength(12);
-          let dateStringToCompareWith = "INITIAL STRING FOR COMPARING DATES";
+          expect(articles).toBeSorted({ key: "created_at", descending: true });
```
Comment (dentednerd) [inline __tests__/app.test.js:195]: Much better!

File: models/articles.model.js [hunk 7]
```diff
@@ -60,23 +41,53 @@ exports.pullArticlesWithCommentCount = () => {
...
       } else {
-        console.table(results.rows);
-        return results.rows;
+        return result.rows[0];
       }
     });
 };
+
+exports.pullArticleComments = (article_id) => {
+  //return pullArticleById(article_id)
```
Comment (dentednerd) [inline models/articles.model.js]: You won't need this comment anymore.

File: models/articles.model.js [hunk 8]
```diff
@@ -60,23 +41,53 @@ exports.pullArticlesWithCommentCount = () => {
...
+          `SELECT comment_id, votes, created_at, author, body
+            FROM comments 
+            WHERE article_id = $1;`,
+          [article_id]
+        )
+        .then((result) => {
+          if (!result.rows.length) {
+            return Promise.reject({
+              status: 204,
+              msg: "NO MESSAGE IS PASSED FOR 200+ ERRORS (NO BODY)",
```
Comment (dentednerd) [inline models/articles.model.js]: 200 statuses aren't error statuses. 204 No Content is still a success status, so Promise.reject isn't appropriate here.

**Review by dentednerd — COMMENTED** [review]
A few changes left to make!

File: models/articles.model.js [hunk 9]
```diff
@@ -60,23 +41,52 @@ exports.pullArticlesWithCommentCount = () => {
...
+    })
+    .then((result) => {
+      return db
+        .query(
+          `SELECT comment_id, votes, created_at, author, body
+            FROM comments 
+            WHERE article_id = $1;`,
+          [article_id]
+        )
+        .then((result) => {
```
Comment (xmenbteam) [inline models/articles.model.js]: This needs to be two separate functions. This function should get all the comments relating to an article, but not the article itself. That's for getARticleById to do!

**Review by xmenbteam — CHANGES_REQUESTED** [review]
Speak after lunch!

File: __tests__/app.test.js [hunk 10]
```diff
@@ -42,15 +35,15 @@ describe("Topics", () => {
         .get("/api/topics")
         .expect(200)
         .then((results) => {
-          expect(typeof results).toBe("object");
+          expect(typeof results.body.topics).toBe("object");
         });
```
Comment (xmenbteam) [inline __tests__/app.test.js:39]: Topics should be an array... How do we test for this?

File: controllers/article.controllers.js [hunk 11]
```diff
@@ -3,15 +3,14 @@ const {
   updateArticleId,
   pullArticlesWithCommentCount,
   pullArticleComments,
+  pushArticleComments,
```
Comment (xmenbteam) [inline controllers/article.controllers.js]: Confusing names for your functions - have a look at my naming conventions document please!

File: models/articles.model.js [hunk 12]
```diff
@@ -5,13 +5,40 @@ exports.pullArticleById = (article_id) => {
     .query(`SELECT * FROM articles WHERE article_id = $1;`, [article_id])
```
Comment (xmenbteam) [inline models/articles.model.js]: This needs to have all the comment_count stuff in it as well!

File: models/articles.model.js [hunk 13]
```diff
@@ -31,52 +58,32 @@ exports.updateArticleId = (article_id, body) => {
     });
 };
 
-exports.pullArticlesWithCommentCount = () => {
+exports.pullArticleComments = (article_id) => {
```
Comment (xmenbteam) [inline models/articles.model.js]: This function is using the comments table, so should be in comments.model.js

File: models/articles.model.js [hunk 14]
```diff
@@ -31,52 +58,32 @@ exports.updateArticleId = (article_id, body) => {
...
+      [article_id]
     )
     .then((result) => {
       return result.rows;
     });
 };
 
-exports.pullArticleComments = (article_id) => {
+exports.pushArticleComments = (article_id, username, body) => {
   return db
```
Comment (xmenbteam) [inline models/articles.model.js]: Again, should be in comments.model.js

**Review by xmenbteam — CHANGES_REQUESTED** [review]
Nice! Just a few bits :)

File: __tests__/app.test.js (hunk 10 above)
Comment (HeatherEver) [reply __tests__/app.test.js:39]: This is still isn't testing for an array specifically, have a look into .toBeArray(

File: __tests__/app.test.js [hunk 15]
```diff
@@ -127,24 +131,40 @@ describe("Articles", () => {
...
+              created_at: expect.any(String),
+              author: expect.any(String),
+              body: expect.any(String),
+            });
+          });
         });
     });
+    test("200 article_id is valid and no comments exist", () => {
+      return request(app).get("/api/articles/2/comments").expect(200);
+    });
```
Comment (HeatherEver) [inline __tests__/app.test.js:167]: need to add a couple of tests for errors here
think about incorrect article id, non existent and invalid

File: __tests__/app.test.js [hunk 16]
```diff
@@ -176,46 +194,34 @@ describe("Articles", () => {
...
         .then((result) => {
-          expect(result.body.msg).toBe("400 - Invalid Request");
+          expect(result.body.restaurant).toMatchObject({
+            comment_id: 19,
+            body: "Thats not news",
+            article_id: 4,
+            author: "lurker",
+            votes: 0,
+            created_at: expect.any(String),
+          });
```
Comment (HeatherEver) [inline __tests__/app.test.js:230]: Need some error testing for POST comment
invalid article id 
invalid comment body
invalid comment username
incorrect path

File: __tests__/app.test.js [hunk 17]
```diff
@@ -176,46 +194,34 @@ describe("Articles", () => {
...
 
 describe("Users", () => {
   describe("GET /api/users", () => {
     test("200 - returns all users", async () => {
       const result = await request(app).get("/api/users").expect(200);
-      const users = result.body;
+      const users = result.body.users;
       expect(Array.isArray(users)).toBe(true);
       expect(users).toHaveLength(4);
       users.forEach((user) => {
```
Comment (HeatherEver) [inline __tests__/app.test.js:293]: Need an error test for this too for incorrect path

**Review by HeatherEver — COMMENTED** [review]
Looks good Mark, just make sure you're testing for errors 😄

File: models/comments.models.js [hunk 18]
```diff
@@ -1,6 +1,7 @@
+const { user } = require("pg/lib/defaults");
 const db = require("../db/connection.js");
 
-exports.fetchArticleComments = (article_id) => {
+exports.fetchArticleCommentsSIMPLE = (article_id) => {
```
Comment (P-Copley) [inline models/comments.models.js]: Do you use this function or is it left over from working things out?

**Pull request merged**